proxy-socket-udp/src/main/java/net/airvantage/proxysocket/udp/ProxyDatagramSocket.java

```diff
@@ -0,0 +1,102 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.function.Predicate;
+
+/**
+ * DatagramSocket that strips Proxy Protocol v2 headers and exposes real client address.
+ *
+ * Thread-safety: This class is thread-safe to the extent that {@link DatagramSocket}
+ * is documented as thread-safe for concurrent send/receive by the JDK. The internal
+ * cache and metrics listener are expected to be thread-safe. The implementation does
+ * not mutate shared state beyond those collaborators.
+ */
+public class ProxyDatagramSocket extends DatagramSocket {
+    private static final Logger LOG = Logger.getLogger(ProxyDatagramSocket.class.getName());
+
+    private ProxyAddressCache addressCache;
+    private ProxyProtocolMetricsListener metrics;
+    private Predicate<InetSocketAddress> trustedProxyPredicate;
+
+    public ProxyDatagramSocket() throws SocketException {
+        super();
+    }
+
+    public ProxyDatagramSocket(SocketAddress bindaddr) throws SocketException {
+        super(bindaddr);
+    }
+
+    public ProxyDatagramSocket(int port) throws SocketException {
+        super(port);
+    }
+
+    public ProxyDatagramSocket(int port, java.net.InetAddress laddr) throws SocketException {
+        super(port, laddr);
+    }
+
+    public ProxyDatagramSocket setCache(ProxyAddressCache cache) { this.addressCache = cache; return this; }
+    public ProxyDatagramSocket setMetrics(ProxyProtocolMetricsListener metrics) { this.metrics = metrics; return this; }
+    public ProxyDatagramSocket setTrustedProxy(Predicate<InetSocketAddress> predicate) { this.trustedProxyPredicate = predicate; return this; }
+
+    @Override
+    public void receive(DatagramPacket packet) throws IOException {
+        super.receive(packet);
+        try {
+            InetSocketAddress lbAddress = (InetSocketAddress) packet.getSocketAddress();
+            if (trustedProxyPredicate != null && !trustedProxyPredicate.test(lbAddress)) {
+                // Untrusted source: do not parse, deliver original packet
+                return;
+            }
+
+            ProxyHeader header = ProxyProtocolV2Decoder.parse(packet.getData(), packet.getOffset(), packet.getLength());
+            if (metrics != null) metrics.onHeaderParsed(header);
+            if (header.isLocal()) {
+                // LOCAL: not proxied
+            } else if (header.isProxy() && header.getProtocol() == ProxyHeader.TransportProtocol.DGRAM) {
+                InetSocketAddress realClient = header.getSourceAddress();
+                if (realClient != null && lbAddress != null) {
+                    if (addressCache != null) addressCache.put(realClient, lbAddress);
+                    packet.setSocketAddress(realClient);
+                }
+            }
+            int headerLen = header.getHeaderLength();
+            packet.setData(packet.getData(), packet.getOffset() + headerLen, packet.getLength() - headerLen);
+        } catch (Exception e) {
```

Comment:
Is catching exception too large ? (I mean should we catch specific exception instead ? )